### PR TITLE
Finish auto-scaler controller

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve? 
ref #1651 
Finish the rest logic of auto-scaler controller to update the `tidbcluster` and `tidbclusterAutoScaler` .
I test the auto-scaling locally with cpu load by sysbench, and it works well.
And I also find some bugs and fix it.

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
Support TidbCluster Auto-scaling feature based by CPU averageUtilization load.
```

### CODE REVIEW

File: pkg/autoscaler/autoscaler/tidb_autoscaler.go [hunk 1]
```diff
@@ -70,7 +70,7 @@ func syncTiDBAfterCalculated(tc *v1alpha1.TidbCluster, tac *v1alpha1.TidbCluster
 }
 
 func updateTcTiDBAnnIfScale(tac *v1alpha1.TidbClusterAutoScaler) {
-	tac.Annotations[label.AnnTiDBLastAutoScalingTimestamp] = time.Now().String()
+	tac.Annotations[label.AnnTiDBLastAutoScalingTimestamp] = fmt.Sprintf("%d", time.Now().Unix())
```
Comment (Yisaer) [inline pkg/autoscaler/autoscaler/tidb_autoscaler.go]: fix record timestamp bug

File: pkg/autoscaler/autoscaler/tikv_autoscaler.go [hunk 2]
```diff
@@ -84,7 +84,7 @@ func filterTiKVInstances(tc *v1alpha1.TidbCluster) []string {
 }
 
 func updateTcTiKVAnnIfScale(tac *v1alpha1.TidbClusterAutoScaler) {
-	tac.Annotations[label.AnnTiKVLastAutoScalingTimestamp] = time.Now().String()
+	tac.Annotations[label.AnnTiKVLastAutoScalingTimestamp] = fmt.Sprintf("%d", time.Now().Unix())
```
Comment (Yisaer) [inline pkg/autoscaler/autoscaler/tikv_autoscaler.go]: fix record timestamp bug

File: pkg/autoscaler/autoscaler/autoscaler_manager.go [hunk 3]
```diff
@@ -55,52 +68,85 @@ func (am *autoScalerManager) Sync(tac *v1alpha1.TidbClusterAutoScaler) error {
...
+		klog.Errorf("tac[%s/%s] tidb sync failed, continue to sync next, err:%v", tac.Namespace, tac.Name, err)
 	}
 	klog.Infof("tc[%s/%s]'s tac[%s/%s] synced", tc.Namespace, tc.Name, tac.Namespace, tac.Name)
 	return nil
 }
 
-//TODO: sync TidbCluster.Spec.Replicas
-func (am *autoScalerManager) syncTidbClusterReplicas(tc *v1alpha1.TidbCluster, oldTCSpec *v1alpha1.TidbClusterSpec) error {
+func (am *autoScalerManager) syncTidbClusterReplicas(tc *v1alpha1.TidbCluster, oldTc *v1alpha1.TidbCluster) error {
+	if apiequality.Semantic.DeepEqual(tc, oldTc) {
```
Comment (DanielZhangQD) [inline pkg/autoscaler/autoscaler/autoscaler_manager.go]: Just compare the whole tc object?

Comment (Yisaer) [reply pkg/autoscaler/autoscaler/autoscaler_manager.go]: Any advice?

File: pkg/autoscaler/autoscaler/autoscaler_manager.go [hunk 4]
```diff
@@ -55,52 +68,85 @@ func (am *autoScalerManager) Sync(tac *v1alpha1.TidbClusterAutoScaler) error {
...
 	defaultTAC(tac)
-	if err := am.syncTiKV(tc, tac, client); err != nil {
-		return err
+	if err := am.syncTiKV(tc, tac, c); err != nil {
+		klog.Errorf("tac[%s/%s] tikv sync failed, continue to sync next, err:%v", tac.Namespace, tac.Name, err)
 	}
-	if err := am.syncTiDB(tc, tac, client); err != nil {
-		return err
+	if err := am.syncTiDB(tc, tac, c); err != nil {
+		klog.Errorf("tac[%s/%s] tidb sync failed, continue to sync next, err:%v", tac.Namespace, tac.Name, err)
```
Comment (DanielZhangQD) [inline pkg/autoscaler/autoscaler/autoscaler_manager.go]: Still have to check if any error occurred, should return error in the end to re-queue.

Comment (DanielZhangQD) [reply pkg/autoscaler/autoscaler/autoscaler_manager.go]: The caller function also has to handle this.

File: pkg/autoscaler/autoscaler/autoscaler_manager.go [hunk 5]
```diff
@@ -55,52 +68,85 @@ func (am *autoScalerManager) Sync(tac *v1alpha1.TidbClusterAutoScaler) error {
...
+		_, updateErr = am.cli.PingcapV1alpha1().TidbClusterAutoScalers(ns).Update(tac)
+		if updateErr == nil {
+			klog.Infof("TidbClusterAutoScaler: [%s/%s] updated successfully", ns, tacName)
+			return nil
+		}
+		klog.Errorf("failed to update TidbClusterAutoScaler: [%s/%s], error: %v", ns, tacName, updateErr)
+		if updated, err := am.taLister.TidbClusterAutoScalers(ns).Get(tacName); err == nil {
+			// make a copy so we don't mutate the shared cache
+			tac = updated.DeepCopy()
+			tac.Annotations = oldTac.Annotations
```
Comment (DanielZhangQD) [inline pkg/autoscaler/autoscaler/autoscaler_manager.go]: status will be synced later?

Comment (Yisaer) [reply pkg/autoscaler/autoscaler/autoscaler_manager.go]: syncing status is in the future plan.

File: pkg/autoscaler/autoscaler/autoscaler_manager.go (hunk 3 above)
Comment (Yisaer) [reply pkg/autoscaler/autoscaler/autoscaler_manager.go]: comparing method updated.

File: pkg/autoscaler/autoscaler/autoscaler_manager.go [hunk 6]
```diff
@@ -97,18 +96,22 @@ func (am *autoScalerManager) syncAutoScaling(tc *v1alpha1.TidbCluster, tac *v1al
 		return err
 	}
 	defaultTAC(tac)
+	oldTikvReplicas := tc.Spec.TiKV.Replicas
 	if err := am.syncTiKV(tc, tac, c); err != nil {
+		tc.Spec.TiKV.Replicas = oldTikvReplicas
```
Comment (DanielZhangQD) [inline pkg/autoscaler/autoscaler/autoscaler_manager.go]: Actually, when error returned, the replica in tc is not updated.

**Pull request merged**